### PR TITLE
feat: implement `error-log-logger` plug-in

### LINKED ISSUE #2410: feat: plugin `error-log-logger`, collect error log and report to remote server by `cosocket`
all of the error log will be written into local files now.

```
$ ls logs
access.log  error.log
```

we can implement a new plugin which named `error-log-logger`, it'll collect the error log from API: https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/errlog.md

we can report those error log to `syslog-ng` server or HTTP server.

### PR DESCRIPTION
**Pull request by dabue**
### What this PR does / why we need it:

add a plugin to collect error log and report to remote TCP server .

fix #2410

you can set the plugin's config in config-default.yaml, likes below:
plugins:                          # plugin list
  - error-log-logger

and  update it links below:
step 1: update the attributes of the plugin

curl http://127.0.0.1:9080//apisix/admin/plugin_metadata/error-log-logger -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
  "host": "127.0.0.1",
  "port": 1999,
  "inactive_timeout": 1
}'

### CODE REVIEW

File: apisix/plugins/error-log-logger.lua [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+        end
+    end
+
+    table.insert(data, "\n")
+    local bytes, err = sock:send(data)
+    if not bytes then
+        sock:close()
+        err_msg = "failed to send data to TCP server: host[" .. config.host
+                  .. "] port[" .. tostring(config.port) .. "] err: " .. err
+        return res, err_msg
```
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:118]: ditto

File: apisix/plugins/error-log-logger.lua [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    end
+
+    sock:setkeepalive(config.keepalive * 1000)
+    return true
+end
+
+
+local function process()
+    if not loaded_flag then
+        local code, body = plugin_metadata.get(plugin_name)
```
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:128]: this way is wrong, please take a look at: https://github.com/apache/apisix/blob/master/apisix/plugins/http-logger.lua#L166

File: apisix/plugins/error-log-logger.lua [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        logs = errlog.get_logs(10)
+    end
+
+    if #entries == 0 then
+        return
+    end
+
+    local log_buffer = buffers[id]
+    if log_buffer then
+        for i = 1, #entries do
```
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:168]: `for i,v in ipairs(table_name) do` this style is better for this case

File: apisix/plugins/error-log-logger.lua [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    local err
+    log_buffer, err = batch_processor:new(send_to_server, config_bat)
+
+    if not log_buffer then
+        core.log.error("error when creating the batch processor: ", err)
+        return
+    end
+
+    buffers[id] = log_buffer
+    for i = 1, #entries do
```
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:192]: ditto

File: apisix/plugins/error-log-logger.lua [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    end
+
+    local err
+    _M.timer, err = core.timer.new("error-log-logger", process)
+    if not _M.timer then
+        core.log.error("failed to create timer error-log-logger: ", err)
+    else
+        core.log.notice("succeed to create timer: error-log-logger")
+    end
+
```
Comment (membphis) [inline apisix/plugins/error-log-logger.lua:217]: remove this line

File: apisix/plugins/error-log-logger.lua [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    ERROR  =    ngx.ERR,
+    WARN   =    ngx.WARN,
+    NOTICE =     ngx.NOTICE,
+    INFO   =    ngx.INFO,
+    DEBUG  =    ngx.DEBUG
+}
+
+
+local config = {
+}
```
Comment (Firstsawyou) [inline apisix/plugins/error-log-logger.lua:68]: I think this `local _M = {}` format will look better, what do you think?

Comment (dabue) [reply apisix/plugins/error-log-logger.lua:68]: fixed it

**Comment by dabue** [comment]
@membphis  please review.

**Comment by membphis** [comment]
@dabue we can merge this PR after https://github.com/apache/apisix/pull/2596 merged.

I think this plugin depends on this PR.

**Comment by dabue** [comment]
@membphis  agree it. After  #2596 merged, I can use the backgroud timer to update this pr.

**Comment by dabue** [comment]
@membphis  this pr has been updated after  #2596  and used the new timer, please have a review.

**Comment by membphis** [comment]
> @membphis this pr has been updated after #2596 and used the new timer, please have a review.

got it, let me take a look ^_^

File: doc/plugins/error-log-logger.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+### Enable plugin
+
+Enable the plug-in `error-log-logger` in `conf/config.yaml`, then this plugin can work fine.
+It does not need to be bound in any route or service.
+
+Here is an example of `conf/config.yaml`:
+
+```yaml
+plugins:                          # plugin list
+  - error-log-logger
```
Comment (membphis) [inline doc/plugins/error-log-logger.md:70]:
```yaml
plugins:                          # plugin list
  ... ...
  - request-id
  - hmac-auth
  - api-breaker
  - error-log-logger              # enable plugin `error-log-logger`

File: doc/plugins/error-log-logger.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+```
+
+### Disable plugin
+
+Remove the plugin `error-log-logger` from `conf/config.yaml`.
```
Comment (membphis) [inline doc/plugins/error-log-logger.md:75]: add a example

File: doc/plugins/error-log-logger.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+## How To Update
+
+step: update the attributes of the plugin
```
Comment (membphis) [inline doc/plugins/error-log-logger.md:79]: `Step`

File: doc/plugins/error-log-logger.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+```shell
+
```
Comment (membphis) [inline doc/plugins/error-log-logger.md:82]: remove this line

File: doc/plugins/error-log-logger.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+```yaml
+plugins:                          # plugin list
+  - error-log-logger
+```
+
+### Disable plugin
+
+Remove the plugin `error-log-logger` from `conf/config.yaml`.
+
+## How To Update
```
Comment (membphis) [inline doc/plugins/error-log-logger.md:77]: bad title, 'How to set the TCP server address'

File: apisix/cli/ngx_tpl.lua [hunk 12]
```diff
@@ -154,6 +154,10 @@ http {
     }
     {% end %}
 
+    {% if enabled_plugins["error-log-logger"] then %}
+        lua_capture_error_log  1m;
```
Comment (membphis) [inline apisix/cli/ngx_tpl.lua:158]: `1m` is small, please change to a larger size.

File: doc/plugins/error-log-logger.md (hunk 7 above)
Comment (dabue) [reply doc/plugins/error-log-logger.md]: updated

File: apisix/cli/ngx_tpl.lua (hunk 12 above)
Comment (dabue) [reply apisix/cli/ngx_tpl.lua]: change to 10m

**Review by membphis — APPROVED** [review]
LGTM

**Comment by membphis** [comment]
@dabue for the doc thing, we need to update the doc index, like this issue: https://github.com/apache/apisix/issues/2790

We can also do this after this PR merge.

File: doc/plugins/error-log-logger.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+- [**How To Enable And Disable**](#how-to-enable-and-disable)
+- [**How To Update**](#how-to-update)
+
+## Name
+
+`error-log-logger` is a plugin which push the log data of APISIX's error.log to TCP servers.
+
+This will provide the ability to send the log data which selected by the level to Monitoring tools and other TCP servers.
+
+This plugin provides the ability to push the log data as a batch to you're external TCP servers. In case if you did not recieve the log data don't worry give it some time it will automatically send the logs after the timer function expires in our Batch Processor.
```
Comment (spacewander) [inline doc/plugins/error-log-logger.md:35]: `you're` => your?
`recieve` => `receive`

File: apisix/plugins/error-log-logger.lua [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+        if not config then
+            core.log.warn("set log filter failed for ", err)
+            return
+        end
+
+    end
+
+    local id = ngx.worker.id()
+    local entries = {}
+    local logs = errlog.get_logs(10)
```
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:148]: Better to get 3's multiple? Like `get_logs(9)`. And we should reuse `logs` as the second argument of `get_logs`.

File: apisix/plugins/error-log-logger.lua [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    if not sock then
+        return false, "failed to init the socket " .. soc_err
+    end
+
+    sock:settimeout(config.timeout * 1000)
+
+    local ok, err = sock:connect(config.host, config.port)
+    if not ok then
+        return false, "failed to connect the TCP server: host[" .. config.host
+                         .. "] port[" .. tostring(config.port) .. "] err: " .. err
```
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:94]: Style: indent

File: apisix/plugins/error-log-logger.lua [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+local function process()
+    local metadata = plugin.plugin_metadata(plugin_name)
+    if not (metadata and metadata.value and metadata.modifiedIndex) then
+        core.log.info("please set the correct plugin_metadata for ", plugin_name)
+        return
+    else
+        local err
+        config, err = lrucache(plugin_name, metadata.modifiedIndex, update_filter,
+                                 metadata.value)
```
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:138]: Style: indent

File: doc/plugins/error-log-logger.md [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+| ---------------- | ------- | ----------- | ------- | ------- | ---------------------------------------------------------------------------------------- |
+| host             | string  | required    |         |         | IP address or the Hostname of the TCP server.                                            |
+| port             | integer | required    |         | [0,...] | Target upstream port.                                                                    |
+| timeout          | integer | optional    | 3       | [1,...] | Timeout for the upstream to connect and send, unit: second.                                                   |
+| keepalive        | integer | optional    | 30      | [1,...] | Time for keeping the cosocket alive, unit: second.                                                   |
+| level            | string  | optional    | WARN    |         | The filter's log level, default warn, chose the level in ["STDERR", "EMERG", "ALERT", "CRIT", "ERR", "ERROR", "WARN", "NOTICE", "INFO", "DEBUG"], the value ERR equals ERROR.         |
```
Comment (spacewander) [inline doc/plugins/error-log-logger.md:48]: "chose" => "choose"

File: apisix/plugins/error-log-logger.lua [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+    type = "object",
+    properties = {
+        host = {type = "string"},
+        port = {type = "integer", minimum = 0},
+        tls = {type = "boolean", default = false},
+        tls_options = {type = "string"},
+        timeout = {type = "integer", minimum = 1, default = 3},
+        keepalive = {type = "integer", minimum = 1, default = 30},
+        name = {type = "string", default = plugin_name},
+        level = {type = "string", default = "WARN"},
```
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:45]: Need to validate with enum, like this: ` enum = {"GET", "POST", "PUT", "DELETE", "PATCH",`

File: apisix/plugins/error-log-logger.lua [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+local ipairs  = ipairs
+local lrucache = core.lrucache.new({
+    ttl = 300, count = 32
+})
+
+
+local metadata_schema = {
+    type = "object",
+    properties = {
+        host = {type = "string"},
```
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:38]: Need to validate with `core.schema.host_def`

**Review by spacewander — CHANGES_REQUESTED** [review]

File: apisix/plugins/error-log-logger.lua (hunk 14 above)
Comment (dabue) [reply apisix/plugins/error-log-logger.lua:148]: LGTM, I'll update it.

File: doc/plugins/error-log-logger.md (hunk 13 above)
Comment (dabue) [reply doc/plugins/error-log-logger.md:35]: :), fix it

File: apisix/plugins/error-log-logger.lua [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+
+    local ok, err = sock:connect(config.host, config.port)
+    if not ok then
+        return false, "failed to connect the TCP server: host[" .. config.host
+                         .. "] port[" .. tostring(config.port) .. "] err: " .. err
+    end
+
+    if config.tls then
+        ok, err = sock:sslhandshake(true, config.tls_options, false)
+        if not ok then
```
[the remark this answers is not in the document]
Comment (dabue) [reply apisix/plugins/error-log-logger.lua:99]: LGTM, I'll fix it.

File: apisix/plugins/error-log-logger.lua (hunk 19 above)
Comment (dabue) [reply apisix/plugins/error-log-logger.lua:38]: that's better.

**Comment by dabue** [comment]
@membphis @spacewander  please review again, thx.

File: .travis/apisix_cli_test.sh.bak [hunk 21]
```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env bash
```
Comment (spacewander) [inline .travis/apisix_cli_test.sh.bak:1]: This file is useless?

File: doc/plugins/error-log-logger.md [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+# Summary
+
+- [**Name**](#name)
+- [**Attributes**](#attributes)
+- [**How To Enable And Disable**](#how-to-enable-and-disable)
+- [**How To Update**](#how-to-update)
+
+## Name
+
+`error-log-logger` is a plugin which push the log data of APISIX's error.log to TCP servers.
```
Comment (spacewander) [inline doc/plugins/error-log-logger.md:31]: push => pushes

File: doc/plugins/error-log-logger.md [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+| keepalive        | integer | optional    | 30      | [1,...] | Time for keeping the cosocket alive, unit: second.                                                   |
+| level            | string  | optional    | WARN    |         | The filter's log level, default warn, choose the level in ["STDERR", "EMERG", "ALERT", "CRIT", "ERR", "ERROR", "WARN", "NOTICE", "INFO", "DEBUG"], the value ERR equals ERROR.         |
+| tls              | boolean | optional    | false   |         | Control whether to perform SSL verification                                              |
+| tls_options      | string  | optional    |         |         | tls options                                                                              |
+| batch_max_size   | integer | optional    | 1000    | [1,...] | Max size of each batch                                                                   |
+| inactive_timeout | integer | optional    | 5       | [1,...] | Maximum age in seconds when the buffer will be flushed if inactive                       |
```
Comment (spacewander) [inline doc/plugins/error-log-logger.md:52]: The `inactive_timeout` in code is 3, but it is 5 in the doc?

File: apisix/plugins/error-log-logger.lua [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+    sock:settimeout(config.timeout * 1000)
+
+    local ok, err = sock:connect(config.host, config.port)
+    if not ok then
+        return false, "failed to connect the TCP server: host[" .. config.host
+            .. "] port[" .. tostring(config.port) .. "] err: " .. err
+    end
+
+    if config.tls then
+        ok, err = sock:sslhandshake(true, config.tls_options, false)
```
Comment (spacewander) [inline apisix/plugins/error-log-logger.lua:100]: The name `tls_options` is bad, in fact it isn't related with any options. Would be better to use `tls_sni` or `tls_server_name`.

File: doc/plugins/error-log-logger.md [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+| port             | integer | required    |         | [0,...] | Target upstream port.                                                                    |
+| timeout          | integer | optional    | 3       | [1,...] | Timeout for the upstream to connect and send, unit: second.                                                   |
+| keepalive        | integer | optional    | 30      | [1,...] | Time for keeping the cosocket alive, unit: second.                                                   |
+| level            | string  | optional    | WARN    |         | The filter's log level, default warn, choose the level in ["STDERR", "EMERG", "ALERT", "CRIT", "ERR", "ERROR", "WARN", "NOTICE", "INFO", "DEBUG"], the value ERR equals ERROR.         |
+| tls              | boolean | optional    | false   |         | Control whether to perform SSL verification                                              |
+| tls_options      | string  | optional    |         |         | tls options                                                                              |
```
Comment (spacewander) [inline doc/plugins/error-log-logger.md:50]: The name tls_options is bad

**Review by spacewander — CHANGES_REQUESTED** [review]

File: .travis/apisix_cli_test.sh.bak (hunk 21 above)
Comment (dabue) [reply .travis/apisix_cli_test.sh.bak:1]: It maybe imported by the rebase, I'll clean it.

File: apisix/plugins/error-log-logger.lua (hunk 24 above)
Comment (dabue) [reply apisix/plugins/error-log-logger.lua:100]: got it

**Comment by membphis** [comment]
@dabue you need to rebase your branch

**Comment by dabue** [comment]
@membphis @spacewander  please review again, thx.

**Review by spacewander — APPROVED** [review]

**Comment by membphis** [comment]
@dabue many thx for your nice job

**Pull request merged**